modules/expandable-panel/ExpandablePanel.tsx

```diff
@@ -0,0 +1,229 @@
+    transform: "rotateZ(0deg)",
+    "@selectors": {
+      "&.expanded": {
+        transform: "rotateZ(-180deg)"
+      }
+    }
+  },
+  headerTitle: {},
+  panel: {
+    overflow: "hidden"
+  }
+});
+
+/**
+ * The ExpandablePanel has a built-in sliding animation when it is expanded/collapsed.
+ *
+ * To be able to achieve this, the body of the panel must have a fixed height so the css
+ * transition works properly.
+ * Since the contents of the panel body are dynamic, the body height will be too.
+ * Thus, before the animation can be triggered, the body height must be calculated.
+ *
+ * What happens when a panel is expanded the first time is the following:
+ * 1. panel body is rendered in the DOM. It is hidden with overflow: hidden, and with the panel's maxHeight
+ *    being set to the panel header's height.
+ * 2. Once the panel body height is calculated, it is set to 0 so the animation can trigger and the height
+ *    can expand from 0 to the actual height. The panel's maxHeight is reset.
+ * 3. Immediately after, the body height is set to its proper value and the animation triggers.
+ *
+ * If the panel is initially expanded already, the body height can be immediately calculated and step 1. is skipped.
+ */
+const ExpandablePanel: React.FC<IProps & WithStylesProps> = ({
+  children,
+  className,
+  cx,
+  header,
+  disabled,
+  expanded,
+  initialExpanded,
+  onClick,
+  styles
+}) => {
+  const windowSize = useWindowSize();
+  const [isExpanded, setIsExpanded] = useState(initialExpanded || false);
+  const [isCollapsing, setIsCollapsing] = useState(false);
+  const [headerHeight, setHeaderHeight] = useState<number | undefined>(0);
+  const [bodyHeight, setBodyHeight] = useState<number | undefined>(0);
+  const [currentBodyHeight, setCurrentBodyHeight] = useState<
+    number | undefined
+  >(0);
+  const headerRef = useRef<HTMLDivElement>(null);
+  const bodyWrapperRef = useRef<HTMLDivElement>(null);
+  const bodyRef = useRef<HTMLDivElement>(null);
+
+  // calculate header height (once or every time window size changes)
+  useEffect(() => {
+    setHeaderHeight(headerRef.current?.getBoundingClientRect().height);
+  }, [headerRef, windowSize]);
+
+  // calculate body height (once or every time window size changes)
+  useEffect(() => {
+    if (bodyHeight === 0 || bodyHeight === undefined) {
+      setBodyHeight(bodyRef.current?.getBoundingClientRect().height);
+    }
+  }, [bodyHeight, bodyRef, isExpanded, windowSize]);
+
+  const handleCollapse = useCallback(() => {
+    setCurrentBodyHeight(0);
+    setIsCollapsing(true);
+
+    // allow the animation to finish before removing the body from the DOM
+    setTimeout(() => {
+      setIsCollapsing(false);
+    }, SLIDE_ANIMATION_DURATION_MS);
+  }, []);
+
+  // set body height manually to be able to have the sliding animation
+  useEffect(() => {
+    if (isExpanded && bodyHeight && bodyHeight > 0 && currentBodyHeight === 0) {
+      setTimeout(() => {
```

Comment:
We should clear the timeout when an effect is triggered too fast to avoid leaks and side effects.

```
useEffect(() => {
const timeout = setTimeout(...);
return () => clearTimeout(timeout);
}

Reply:
Absolutely 👍